robot_skills/src/robot_skills/robot_part.py

```diff
@@ -99,10 +109,71 @@ def create_service_client(self, name, srv_type):
         self._add_connection(name, srv)
         return srv
 
+    def create_subscriber(self, name, *args, **kwargs):
+        """
+        Creates a Subscriber and add to the connections to check
+        :param name: string with the name topic to subscribe
+        :param args: other args passed to rospy.Subscriber
+        :param kwargs: other keyword args passed to rospy.Subscriber
+        :return: the service client
+        """
+        sub = rospy.Subscriber(name, *args, **kwargs)
+        self._add_connection(name, sub)
+        return sub
+
     def _add_connection(self, name, connection):
         """
         Adds a connection to the internal dict with connections that is used when initializing the robot object.
         :param name: name of the connection
-        :param connection: connection to add
+        :param connection: connection to add. This might be a ServiceProxy, ActionClient or Subscriber
         """
         self.__ros_connections[name] = connection
+
+    @property
+    def operational(self):
+        """
+        Check whether this bodypart's hardware is operational
+        If the associated hardware is not yet up, has an error etc, the bodypart is not operational
+
+        :rtype: bool True is part is ready to do work, False otherwise
+        """
+        return self._operational
+
+    def start_check_operational(self, name):
+        """
+        Start to check if the bodypart is operational. To do so, subscribe to the hardware status/diagnostics
+
+        :param name: check on the level of the diagnostic_msgs/DiagnosticStatus with this name
+        """
+
+        self._operational = False
+
+        self.__diagnostics_name = name
+
+        self._hardware_status_sub = self.create_subscriber("/" + self.robot_name + "/hardware_status",
```

Comment:
So each body part creates a subscriber to the same topic? That seems a lot of overhead. I would suggest to make this part of the Robot object, and have the callback function relay the relevant part of the message to the various body parts. Another advantage would be that you don't have to call the start_check_operational method in each (relevant) body part. Only thing is that you would have to consider when the check is required